src/lib/udev/google_nvme_id

```diff
@@ -0,0 +1,305 @@
+  if [[ "${dev_path}" =~ $PARTITION_NUMBER_REGEX ]]; then
+    partition_number="${BASH_REMATCH[1]}"
+    echo "${partition_number}"
+  else
+    echo ''
+  fi
+  return 0
+}
+
+#######################################
+# Generates a symlink for a PD-NVMe device using the metadata's disk name.
+# Primarily used for testing but can be used if the script is directly invoked.
+# Globals:
+#   Uses ID_SERIAL_SHORT (can be populated by identify_pd_disk)
+# Arguments:
+#   The device path for the disk
+#######################################
+function gen_symlink() {
+  local dev_path="${1}"
+  local partition_number="$(get_partition_number "${dev_path}")"
+
+  if [[ -n "${partition_number}" ]]; then
+    ln -s "${dev_path}" /dev/disk/by-id/google-persistent-disk-"${ID_SERIAL_SHORT}"-part"${partition_number}" > /dev/null 2>&1
+  else
+    ln -s "${dev_path}" /dev/disk/by-id/google-persistent-disk-"${ID_SERIAL_SHORT}" > /dev/null 2>&1
+  fi
+
+  return 0
+}
+
+#######################################
+# Validates that a given path corresponds to a valid NVMe device
+# Globals:
+#   Uses nvme_cli_bin
+# Arguments:
+#   The device path for the disk
+# Returns:
+#   The return value of the call to the nvme id-ctrl command
+#######################################
+function validate_nvme_device() {
+  "${nvme_cli_bin}" id-ctrl "${1}" &>/dev/null
+}
+
+#######################################
+# Determines if the device path is a PD or Local SSD disk
+# Globals:
+#   Uses nvme_cli_bin, PD, and LOCAL_SSD
+# Arguments:
+#   The device path for the disk
+# Returns:
+#   1 if the device type is Unknown, 0 on success
+#######################################
+function detect_pd_or_local_ssd() {
+  local nvme_id_ctrl
+  nvme_id_ctrl="$("${nvme_cli_bin}" id-ctrl "${1}")"
+  if [[ "${nvme_id_ctrl}" =~ 'nvme_card-pd' ]]; then
+    echo "${PD}"
+  elif [[ "${nvme_id_ctrl}" =~ 'nvme_card' ]]; then
+    echo "${LOCAL_SSD}"
+  else
+    return 1
+  fi
+  return 0
+}
+
+#######################################
+# Populates the ID_* global variables with a disk's device name and namespace
+# Globals:
+#   Populates ID_NAMESPACE_INDEX, ID_SERIAL_SHORT, and ID_SERIAL
+# Arguments:
+#   The device path for the disk
+# Returns:
+#   0 on success and 255 if an error occurrs
+#######################################
+function identify_pd_disk() {
+  local dev_path="${1}"
+  local dev_name
+  dev_name="$(get_namespace_device_name "${dev_path}")"
+  if [[ $? != 0 ]]; then
+    return 255
+  fi
+
+  ID_NAMESPACE_INDEX="${namespace_index}"
+  ID_SERIAL_SHORT="${dev_name}"
+  ID_SERIAL="Google_PersistentDisk_${ID_SERIAL_SHORT}"
+  return 0
+}
+
+#######################################
+# Ensures a given command/path exists
+# Arguments:
+#   The path for the command
+# Returns:
+#   0 if the command is found/executable and 1 on error
+#######################################
+function confirm_command() {
+  local command_to_check="${1}"
+  command -v "${command_to_check}" > /dev/null 2>&1
```

Comment:
this single line can be used in place of this function

Reply:
Done.